### PR TITLE
Post about migrating MySQL Connector to Oracle's ANTLR Grammar

### PR DESCRIPTION
**Pull request by vsantonastaso**

### CODE REVIEW

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+
+=== What exactly is a grammar?
+
+_A grammar is a formal specification of a language's syntax_.
+
+It defines what's valid and what's not. You could try parsing SQL with regular expressions, but that seems faster. SQL has nested structures, context-sensitive keywords, and ambiguous syntax that regex simply can't handle reliably.
+
+In our case, we're talking about _ANTLR_ (ANother Tool for Language Recognition), which is a parser generator. You write grammar rules in a special format (.g4 in our case), and ANTLR generates Java code that can parse text matching those rules.
+
+MySQL Server actually uses _Yacc_ officially for its own parser, with .yy grammar files.So why didn't Debezium just use that?
```
Comment (mfvitale) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
MySQL Server actually uses _Yacc_ officially for its own parser, with .yy grammar files. So why didn't Debezium just use that?
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+*Parser grammar* - Defines how tokens combine into valid statements. This is where the actual syntax rules live:
+
+[source,antlr]
+----
+createTable
+    : CREATE_SYMBOL TABLE_SYMBOL tableName ...
+    ;
+----
+
+MySQL's complete grammar is _massive_. The Oracle grammar we migrated to has over 3,600 lines in the lexer and 5,400+ lines in the parser, at this moment. It covers everything from basic CREATE TABLE to more edge features like window functions, JSON_TABLE, spatial types, and MySQL-specific syntax.
```
Comment (mfvitale) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
MySQL's complete grammar is _massive_. The Oracle grammar we migrated to has over *3,600* lines in the lexer and *5,400+* lines in the parser, at this moment. It covers everything from basic CREATE TABLE to more edge features like window functions, JSON_TABLE, spatial types, and MySQL-specific syntax.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+=== Why we migrated from Positive Technologies to Oracle
+
+Debezium originally used the Positive Technologies MySQL grammar. The main reason we had to move is that the PT grammar has been https://github.com/antlr/grammars-v4/pull/4443[deprecated] in the grammars-v4 repository in favor of Oracle's official grammar. On top of that, it accepted SQL that MySQL would actually reject, for example, it let you write `NVARCHAR` without specifying a length, which MySQL doesn't allow. That meant Debezium could parse invalid DDL and build an incorrect schema without anyone noticing.
+
+Meanwhile, Oracle. who owns MySQL, published an official ANTLR grammar for MySQL. We maintain our own copy in Debezium, but having the official grammar as our reference changed everything. When MySQL adds new syntax or fixes parsing edge cases, we have an authoritative source to sync against. No more guessing. No more reverse-engineering MySQL's parser from documentation and hoping we got it right. When we need to pull upstream changes, we're aligning with the same grammar Oracle uses internally, not some community interpretation that might have missed edge cases.
```
Comment (mfvitale) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]: > Meanwhile, Oracle. who owns MySQL

Maybe `.` should be a `,`?

Just a note, we tend to have a single sentence on one line so it will facilitate the review.

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+=== How ANTLR works in Debezium
+
+When ANTLR processes a .g4 grammar file, it generates several Java classes. For MySQL, you get `MySqlLexer` (tokenizes the input), `MySqlParser` (builds the parse tree), and a few base classes for walking that tree. The question is: how do you actually extract information from the parse tree?
+
+ANTLR gives you two options. The Visitor pattern lets you manually walk the tree node by node, useful when you're transforming code, but more work than we need here. We use the Listener pattern instead. ANTLR walks the tree for you and fires events when it enters and exits each rule. We just hook into the events we care about and extract what we need. Since we're not rewriting SQL or controlling traversal order, listeners are the natural fit, we want to react to what the parser finds, not drive the walk ourselves.
```
Comment (mfvitale) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
ANTLR gives you two options. The `Visitor` pattern lets you manually walk the tree node by node, useful when you're transforming code, but more work than we need here. We use the `Listener` pattern instead. ANTLR walks the tree for you and fires events when it enters and exits each rule. We just hook into the events we care about and extract what we need. Since we're not rewriting SQL or controlling traversal order, listeners are the natural fit, we want to react to what the parser finds, not drive the walk ourselves.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+
+        parser.databaseTables().overwriteTable(tableEditor.create());
+        parser.signalCreateTable(tableEditor.tableId(), ctx.getParent());
+    }, tableEditor);
+
+    tableEditor = null;
+}
+----
+
+It sets the default charset if needed and commits the table to Debezium's in-memory schema model. At that point we have a complete picture of `users`, ready to be used when binlog events come in.
```
Comment (mfvitale) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
It sets the default charset if needed and saves the table to Debezium's in-memory schema model. At that point we have a complete picture of `users`, ready to be used when binlog events come in.
```

Maybe here `commit` can be misleading. WDYT?

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+
+*This is important context for the migration.* When we switched from the Positive Technologies grammar to Oracle's grammar, the generated Java classes changed. Method signatures changed. Parse tree structure changed. The rule for handling default values in the PT grammar was named differently than in Oracle's grammar. Every listener that assumed a specific tree shape or token name had to be updated. That's where most of our pain during the migration.
```
Comment (mfvitale) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
*This is the important context for the migration.* When we switched from the Positive Technologies grammar to Oracle's grammar, the generated Java classes changed. Method signatures changed. Parse tree structure changed. The rule for handling default values in the PT grammar was named differently than in Oracle's grammar. Every listener that assumed a specific tree shape or token name had to be updated. That's where most of our pain during the migration.
```

**Review by mfvitale — APPROVED** [review]
Very nice read!

File: _posts/2026-05-10-mysql-grammar-migration.adoc (hunk 3 above)
Comment (vsantonastaso) [reply _posts/2026-05-10-mysql-grammar-migration.adoc]: yes it's a typo

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+Debezium's MySQL connector reads MySQL's binary log (binlog), a sequential record of every database change.
+It transforms these binlog entries into structured change events and streams them for downstream consumers.
+
+When the MySQL connector reads from the binlog, it analyzes row-level changes (inserts, updates, deletes) and here's the catch: the binlog doesn't include the table schema with each event.
+It only tells you "_row X in table Y changed_" with an array of column values.
+No column names, no types, nothing.
+
+To generate meaningful change events, Debezium needs to know what each column is called and what type it is.
+We can't just ask the database for the current schema because by the time we're reading an event from last week, the table might have been altered.
+And hitting the database for schemas every time would add unnecessary load on production.
```
Comment (vjuranek) [inline _posts/2026-05-10-mysql-grammar-migration.adoc:30]: Maybe mention explicitly (implicitly it's in the next sentence) that another reason why we cannot query DB for the schema is that there can be a lag and we process row which has different schema than the current schema in the database.

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+
+=== What exactly is a grammar?
+
+_A grammar is a formal specification of a language's syntax_.
+
+It defines what's valid and what's not.
+You could try parsing SQL with regular expressions, but that seems faster.
+SQL has nested structures, context-sensitive keywords, and ambiguous syntax that regex simply can't handle reliably.
+
+In our case, we're talking about _ANTLR_ (ANother Tool for Language Recognition), which is a parser generator.
```
Comment (vjuranek) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]: Add link to ANTLR?

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+[source,sql]
+----
+CREATE TABLE metrics (
+    timestamp BIGINT,
+    lag INT,              -- LAG as column name
+    rank INT              -- RANK as column name
+);
+----
+
+The Oracle grammar has different keyword categories:
```
Comment (vjuranek) [inline _posts/2026-05-10-mysql-grammar-migration.adoc:341]: The list renders on a single line. Probably missing empty line here?

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+[source,sql]
+----
+ALTER TABLE t ADD COLUMN (
+    created TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    updated TIME DEFAULT '08:00'
+);
+----
+
+Default values were being assigned to the wrong columns.
+The `created` column ended up with `updated`'s default value.
```
Comment (vjuranek) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]: Some issue with rendering `updated` in italic. No idea what could be wrong here.

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+One unexpected benefit of migrating to the official grammar was that we could actually improve it.
+While running our test suite against the Oracle grammar, I kept hitting cases where valid MySQL syntax was being rejected.
+Not edge cases from old DDL dumps, stuff you'd write today that works fine on MySQL 8.0.46 and 9.0.1 but fails in the grammar.
+
+The first one I noticed was `GEOMCOLLECTION`.
+MySQL has accepted it as a synonym for `GEOMETRYCOLLECTION` for years, in column types, in `CAST`, in spatial functions.
+But the grammar only knew about the long form.
+Then I found that charset introducers like `_UTF8MB4'text'` were rejected because the lexer pattern only matched lowercase `[a-z0-9]+`.
+Real-world DDL has both cases, and MySQL doesn't care.
+Same story with backtick escaping: `` `test``table` `` is how MySQL handles backticks inside identifiers, but the `BACK_TICK_QUOTED_ID` rule didn't recognize the doubled backtick as an escape sequence.
```
Comment (vjuranek) [inline _posts/2026-05-10-mysql-grammar-migration.adoc:452]: Same issue here with backticks

**Review by vjuranek — APPROVED** [review]
Nice blog post @vsantonastaso ! I left few comment mostly about rendering the text. Not sure if it not outside the scope of this blog post, but maybe a short notice how we handle MariaDB grammar would be good?

File: _posts/2026-05-10-mysql-grammar-migration.adoc (hunk 7 above)
Comment (vsantonastaso) [reply _posts/2026-05-10-mysql-grammar-migration.adoc:30]: I added explicitly the lag issue. thanks

**Comment by vsantonastaso** [comment]
> Nice blog post @vsantonastaso ! I left few comment mostly about rendering the text. Not sure if it not outside the scope of this blog post, but maybe a short notice how we handle MariaDB grammar would be good?

Thanks @vjuranek . Addressed rendering issues. Regarding MariaDB grammar, I don’t think it’s worth covering it here, as it’s completely decoupled from MySQL on the implementation side. Mentioning MariaDB would also imply mentioning Oracle grammar, which I don’t think fits well with the context of this blog post. WDYT?

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 12]
```diff
@@ -0,0 +1,491 @@
...
+[source,antlr]
+----
+createTable
+    : CREATE_SYMBOL TABLE_SYMBOL tableName ...
+    ;
+----
+
+MySQL's complete grammar is _massive_.
+The Oracle grammar we migrated to has over *3,600* lines in the lexer and *5,400+* lines in the parser, at this moment.
+It covers everything from basic CREATE TABLE to more edge features like window functions, JSON_TABLE, spatial types, and MySQL-specific syntax.
```
Comment (gunnarmorling) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]: May be worth pointing out that while stuff like window functions aren't relevant from a Debezium perspective, we still have to be able to parse each single bit of the language, i.e. there's no way to only parse the stuff we actually care about.

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 13]
```diff
@@ -0,0 +1,491 @@
...
+Before diving into grammars, let's talk about why Debezium needs a DDL parser.
+
+Debezium's MySQL connector reads MySQL's binary log (binlog), a sequential record of every database change.
+It transforms these binlog entries into structured change events and streams them for downstream consumers.
+
+When the MySQL connector reads from the binlog, it analyzes row-level changes (inserts, updates, deletes) and here's the catch: the binlog doesn't include the table schema with each event.
+It only tells you "_row X in table Y changed_" with an array of column values.
+No column names, no types, nothing.
+
+To generate meaningful change events, Debezium needs to know what each column is called and what type it is.
```
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
To generate meaningful change events, Debezium needs to know the name and type of each column.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 14]
```diff
@@ -0,0 +1,491 @@
...
+We can't just ask the database for the current schema because the connector can lag behind and by the time we're reading an event, the table might have been changed.
```
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
We can't just ask the database for the current schema because the connector can lag behind, and by the time we read an event, the table might have changed.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 15]
```diff
@@ -0,0 +1,491 @@
...
+And hitting the database for schemas every time would add unnecessary load on production.
+We need the schema as it was _at that exact point in time_.
+
+The solution?
+Parse every DDL statement in the binlog and maintain an in-memory model of how each table's schema evolved over time.
+When we encounter an _INSERT_ event, we use the table definition from that moment to generate the right change event.
+
+This is why getting DDL parsing right is critical.
+A bug in parsing means we might miss a column, get the wrong type, or completely fail to process events.
```
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
A bug in parsing can cause us to miss a column, get the wrong type, or completely fail to process events.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 16]
```diff
@@ -0,0 +1,491 @@
...
+_Yacc_ and _Bison_ were designed around C and C++.
+They're fast and reliable, but they generate C code.
+If you want to use them in Java, you're looking at JNI bindings for example or porting the entire grammar manually.
+The syntax is also tightly coupled to C conventions, which makes adaptation harder.
+
+ANTLR, on the other hand, was built with multi-language support from the start.
+The same .g4 grammar file can generate parsers for Java, Python, JavaScript, C#, or whatever target you need.
+The grammar syntax is cleaner and more declarative.
+You can embed Java code directly in the grammar to customize parsing behavior.
+And the tooling ecosystem is significantly better, with IDE plugins, parse tree visualization, and debugging support built in.
```
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
And the tooling ecosystem is significantly better, with IDE plugins, parse-tree visualization, and built-in debugging support.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc (hunk 12 above)
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
It covers everything from basic CREATE TABLE to advanced features such as window functions, JSON_TABLE, spatial types, and MySQL-specific syntax.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 17]
```diff
@@ -0,0 +1,491 @@
...
+createTable
+    : CREATE_SYMBOL TABLE_SYMBOL tableName ...
+    ;
+----
+
+MySQL's complete grammar is _massive_.
+The Oracle grammar we migrated to has over *3,600* lines in the lexer and *5,400+* lines in the parser, at this moment.
+It covers everything from basic CREATE TABLE to more edge features like window functions, JSON_TABLE, spatial types, and MySQL-specific syntax.
+
+The grammar also includes semantic predicates (code that runs during parsing) to handle things like version specific features, context-sensitive keywords, and character set introducers.
```
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
The grammar also includes semantic predicates (code that runs during parsing) to handle version-specific features, context-sensitive keywords, and character set introducers.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 18]
```diff
@@ -0,0 +1,491 @@
...
+
+=== Why we migrated from Positive Technologies to Oracle
+
+Debezium originally used the Positive Technologies MySQL grammar.
+The main reason we had to move is that the PT grammar has been https://github.com/antlr/grammars-v4/pull/4443[deprecated] in the grammars-v4 repository in favor of Oracle's official grammar.
+On top of that, it accepted SQL that MySQL would actually reject, for example, it let you write `NVARCHAR` without specifying a length, which MySQL doesn't allow.
+That meant Debezium could parse invalid DDL and build an incorrect schema without anyone noticing.
+
+Meanwhile, Oracle, who owns MySQL, published an official ANTLR grammar for MySQL.
```
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
Meanwhile, Oracle, which owns MySQL, published an official ANTLR grammar for MySQL.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 19]
```diff
@@ -0,0 +1,491 @@
...
+When we need to pull upstream changes, we're aligning with the same grammar Oracle uses internally, not some community interpretation that might have missed edge cases.
+
+=== How ANTLR works in Debezium
+
+When ANTLR processes a .g4 grammar file, it generates several Java classes.
+For MySQL, you get `MySqlLexer` (tokenizes the input), `MySqlParser` (builds the parse tree), and a few base classes for walking that tree.
+The question is: how do you actually extract information from the parse tree?
+
+ANTLR gives you two options.
+The `Visitor` pattern lets you manually walk the tree node by node, useful when you're transforming code, but more work than we need here.
```
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
The `Visitor` pattern lets you manually walk the tree node by node, which is useful when you're transforming code, but it's more work than we need here.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 20]
```diff
@@ -0,0 +1,491 @@
...
+    listeners.add(new CreateAndAlterDatabaseParserListener(parser));
+    listeners.add(new DropDatabaseParserListener(parser));
+    listeners.add(new CreateTableParserListener(parser, listeners));
+    listeners.add(new AlterTableParserListener(parser, listeners));
+    listeners.add(new DropTableParserListener(parser));
+    listeners.add(new RenameTableParserListener(parser));
+}
+----
+
+When ANTLR walks through the parse tree, it passes events to the listeners that are listening each specific type of statement.
```
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
When ANTLR walks the parse tree, it passes events to listeners for each specific statement type.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 21]
```diff
@@ -0,0 +1,491 @@
...
+            }
+            if (signed.ulong_number() != null) {
+                getColumnEditor().defaultValueExpression(sign + signed.ulong_number().getText());
+            }
+        }
+    }
+}
+----
+
+Notice that `getColumnEditor()` doesn't hold a direct reference, it asks the parent listener for whatever column is currently being processed.
```
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
Notice that `getColumnEditor()` doesn't hold a direct reference, but instead, it asks the parent listener for the column currently being processed.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc [hunk 22]
```diff
@@ -0,0 +1,491 @@
...
+----
+
+It sets the default charset if needed and saves the table to Debezium's in-memory schema model.
+At that point we have a complete picture of `users`, ready to be used when binlog events come in.
+
+*This is the important context for the migration.*
+When we switched from the Positive Technologies grammar to Oracle's grammar, the generated Java classes changed.
+Method signatures changed.
+Parse tree structure changed.
+The rule for handling default values in the PT grammar was named differently than in Oracle's grammar.
```
Comment (Naros) [inline _posts/2026-05-10-mysql-grammar-migration.adoc]:
```suggestion
The rule for handling default values in the PT grammar was named differently from that in Oracle's grammar.
```

File: _posts/2026-05-10-mysql-grammar-migration.adoc (hunk 12 above)
Comment (vsantonastaso) [reply _posts/2026-05-10-mysql-grammar-migration.adoc]: Thanks @gunnarmorling . Mentioned it

**Comment by vsantonastaso** [comment]
Thanks @Naros . Addressed all suggestions

**Review by Naros — APPROVED** [review]

**Comment by Naros** [comment]
@vsantonastaso feel free to merge whenever you're ready to publish.

**Review by kmos — APPROVED** [review]
LGTM 🚀
